core/src/main/java/jeeves/transaction/TransactionManager.java

```diff
@@ -77,25 +80,57 @@
             if (exception[0] == null) {
                 exception[0] = e;
             }
-            rolledBack = true;
-            doRollback(context, transactionManager, transaction);
+
+            try {
+                doRollback(context, transactionManager, transaction);
+            }
+            finally {
+                rolledBack = true;
+            }
         } finally {
             try {
-                if (readOnly) {
-                    doRollback(context, transactionManager, transaction);
+                if (readOnly && !rolledBack) {
+                    try {
+                        doRollback(context, transactionManager, transaction);
+                    }
+                    finally {
+                        rolledBack = true;
+                    }
                 } else if (!rolledBack && (isNewTransaction || commitBehavior == CommitBehavior.ALWAYS_COMMIT)) {
                     doCommit(context, transactionManager, transaction);
                 }
             } catch (TransactionSystemException e) {
                 if (!(e.getOriginalException() instanceof RollbackException)) {
                     Log.error(Log.JEEVES, "ERROR committing transaction, will try to rollback", e);
-                    doRollback(context, transactionManager, transaction);
+                    if (!rolledBack) {
+                        try {
+                            doRollback(context, transactionManager, transaction);
+                        }
+                        finally {
+                            rolledBack = true;
+                        }
+                    }
                 } else {
                     Log.debug(Log.JEEVES, "ERROR committing transaction, will try to rollback", e);
+                    if (!rolledBack) {
+                        try {
+                            doRollback(context, transactionManager, transaction);
+                        }
+                        finally {
+                            rolledBack = true;
```

Comment:
This seem not used later.

Reply:
I did not change any logic here, however I can see a test for `!rollBack` check above.
Reading this code I would assume this flag is used to avoid calling `doRollback` multiple times.

Reply:
Added debug messages to mimic the kind of checks I was performing with the debugger.